ASCII + Latin-1 + Latin Extended A + currency + general punctuation subset data added to lib.plist

chrissimpkins opened the pull request:
…for subset builds

@burodepeper If you have some time, can I ask you to double check me on these lists for the web font subset?  They are all identical new `lib.plist` files across each variant so only necessary to examine one set.  I just want to confirm that another set of eyes has a look through to confirm that I didn't insert a misspelling or skip one of the glyphs in the sets.

The only one that *should* be missing is U+017F and that is because it is not included in our sets.

The list requires the glyph "name" not the file path to the .glif file.

The Unicode Standard is here: http://www.unicode.org/Public/6.2.0/charts/CodeCharts.pdf

Thanks

TODO: 

- [x] how to handle currency symbols and general punctuation glyphs (see https://github.com/source-foundry/Hack/pull/291#discussion_r135722204 and https://github.com/source-foundry/Hack/pull/291#discussion_r135726544)
- [x] build test sets to determine final web font file sizes for woff and woff2

chrissimpkins commented:
had to add the `public.glyphOrder` key back to the lib.plist file. fontmake uses it to determine the glyphs that are built into the font (without subsetting).  The subset list is an array under the key `com.schriftgestaltung.Keep Glyphs`.

burodepeper commented on source/Hack-Regular.ufo/lib.plist:
```diff
@@ -2,3346 +2,324 @@
-		<string>currency</string>
-		<string>dollar</string>
-		<string>dong</string>
-		<string>Euro</string>
```
I think it is a smart idea if we also include the Euro sign, or in general, all the currency signs we have available that are still actively in use. $, £, ¢ and ¥ are already covered I believe by Latin-1 and Latin-1 Supplement. Or perhaps even just the currency sign block, 20A0-20CF.

burodepeper commented on source/Hack-Regular.ufo/lib.plist:
```diff
@@ -2,3346 +2,324 @@
-		<string>uni10F3</string>
-		<string>uni10F4</string>
-		<string>uni10F5</string>
-		
```
315 glyphs counted.

burodepeper replied:
I expected:
- Basic Latin: 94 glyphs
- Latin-1 Supplement: 94 glyphs
- Latin Extended-A: 128 glyphs

For a total of 316 glyphs. With `017F` missing, that seems about right. I haven't checked for typos, but I feel safe to assume that you didn't manually type this list.

burodepeper replied:
If we consider the missing Euro sign, what other obvious things might we have missed?

burodepeper replied:
I think we should also consider adding: General punctuation (2000-206F) for fancy quotes often found in sources that have been created in non text-editor word processors, and also for the en- and em-dash. In my opinion, those characters should be encoded, but you obviously can't rely on that.

chrissimpkins replied:
> 315 glyphs counted.

ty!

chrissimpkins commented:
> but I feel safe to assume that you didn't manually type this list.

and you would be incorrect in that assumption ;)  I did type the individual name strings, not the XML elements.  XCode is a very good property list editor.  I don't have any approach that labels 1:1 glyph names in our sets with the Unicode values.

chrissimpkins commented:
Added the currency symbol / general punctuation suggestions to the TODO list.  It is a balance between what will be used in a monospaced font and the size of the final files.  Users will always have the full sets if they need extended symbols.  Out of interest, do web fonts replace individual glyphs with a fallback font in the CSS if any glyph in the entire formatted body of text is missing in the set, or is it only the individual missing glyph(s) that is replaced?

chrissimpkins commented:
Shall we build some sets and compare final file sizes for woff and woff2 sets?  No idea what kind of overhead there is in the metadata, but with starting sizes for whole sets at ~150kb (woff) and ~100kb (woff2), we should be down in the vicinity of 30kb and 20kb, respectively.  Data that I have is "average" web page size at present is 2MB.  Web fonts take advantage of caching in the browsers so only first page view impact.  Certainly not my area, but to me seems like we have room to add more glyphs if felt to be important.

chrissimpkins commented:
> and you would be incorrect in that assumption

see how much work your suggestions lay on me :)

burodepeper commented:
> It is a balance between what will be used in a monospaced font and the size of the final files.

I think the € is a must-have, as are the fancy quotes. I come across both of those myself on a regular basis. Extending that to their (relatively small) groups seems to cover similar use-cases.

> Out of interest, do web fonts replace individual glyphs with a fallback font in the CSS if any glyph in the entire formatted body of text is missing in the set, or is it only the individual missing glyph(s) that is replaced?

As far as I know, web fonts are treated like any other font, they're just loaded off of the web. That would mean that individual missing glyphs will be replaced by whatever the software or OS uses as fallback, and at that, it will _not_ use a monospaced fallback, if monospace isn't specified in the CSS as fallback in the font-family. We need to check that too.

> Shall we build some sets and compare final file sizes for woff and woff2 sets? No idea what kind of overhead there is in the metadata, but with starting sizes for whole sets at ~150kb (woff) and ~100kb (woff2), we should be down in the vicinity of 30kb and 20kb, respectively.

Sure. At those sizes, the biggest factor is the latency setting up a connection to the server/CDN. And that is a good thing, because then it is no longer 'our problem' ;)

burodepeper commented:
> see how much work your suggestions lay on me :)

Perhaps it is revenge ;)

chrissimpkins commented:
> Perhaps it is revenge ;)

I have no doubt...

chrissimpkins commented:
Will try to build them tonight and will update here when we have more information.  

I'm good with all suggestions about expansion of sets that you had above.

chrissimpkins commented:
@anthrotype Cosimo, I modified our lib.plist files in order to add the Glyphs subset keys (for our web font builds) and during this lib.plist cleanup, I removed the public.glyphOrder key/values.  This caused our fontmake build to fail.

Start line for the fontmake exception:
https://travis-ci.org/source-foundry/Hack/jobs/269423508#L4217

My understanding is that lib.plist is supposed to be an optional file in the UFO spec.  I wanted to see if

1) this lib.plist data is considered mandatory for fontmake builds
2) you might consider converting to contents.plist (mandatory file in UFO spec) for the glyph lists necessary for builds from UFO source

We are struggling with how to manage the editor dependent lib.plist data that is unnecessary in the remote git repository in a collaborative typeface development project.  Ideally, everyone keeps their editor dependent data locally where they need it and builds occur from data in mandatory files that do not rely on the editor in use by the developer who commits.

The changes that occur in these files are not terribly important, but they do require review prior to merge and with thousands of lines that can change it makes for unnecessary work for the source review team.

anthrotype commented:
Right, it should be optional, although the values prefixed with `public.*` would be good to keep.
Do you mind opening an issue on fontmake's repo?

chrissimpkins commented:
Here is where we are with full set vs. current ASCII / Latin-1 / Latin Extended A web fonts

#### Full set

![full set](https://user-images.githubusercontent.com/4249591/29850815-4836b12e-8cfe-11e7-949f-203a0568926d.png)

#### Subset

![subsets](https://user-images.githubusercontent.com/4249591/29850823-4d975a74-8cfe-11e7-894e-c2960eae6ffc.png)

chrissimpkins commented:
For some reason, the use of the lib.plist files with the subset data fields in them broke the ttf builds *without the use of the `--subset` flag* in fontmake.  I hacked together a temporary source directory approach for subsetting that involves a copy of source directories to temp dir, replacement of the lib.plist without subsetting data with the lib.plist that contains subsetting data, and then subset builds.  

Not pretty but it works...  The temporary source directories are cleaned up at the end of the process and recreated with every new subset build.

As of https://github.com/source-foundry/Hack/pull/291/commits/8895e7e9497df92e34e3465e50f409bb2b80644f, you can build each file type individually:

```
$ make ttf
$ make woff
$ make woff2
$ make subsets
```
or build all of them in one fell swoop:

```
$ make
```

All above require previous dependency installs.  If they are not installed yet, you can install all depdencies and build all release files with:

```
$ make build-with-dependencies
```

chrissimpkins commented:
The subsets seem to pull in additional glyphs including the likes of superscripts/subscripts/all fractions, though they are not all specified in the lib.plist subset data.  Appear to be one short in bold and bold italic sets c/w regular and italic sets.  Will need to go through them glyph by glyph to confirm that we are capturing what we intend to in the subsets.

Let me know based upon current file sizes (https://github.com/source-foundry/Hack/pull/291#issuecomment-325848444) whether you still feel that we should add the additional sets that you recommended above.

burodepeper commented:
> Let me know based upon current file sizes (#291 (comment)) whether you still feel that we should add the additional sets that you recommended above.

Sure, those numbers are good!

chrissimpkins commented:
Currency glyphs added to web font subsets https://github.com/source-foundry/Hack/pull/291/commits/869c0d1d8476b78003fb9db708487ccb34617f51

General punctuation glyphs added to web font subsets https://github.com/source-foundry/Hack/pull/291/commits/c750713fe0e5e8d18c95e8802daff05bd776dd79

chrissimpkins commented:
Minimal impact on file sizes with addition of new glyphs to subsets

![web 2017-08-30 21-10-19](https://user-images.githubusercontent.com/4249591/29901745-ab227614-8dc7-11e7-97de-09916b8089f3.png)

chrissimpkins commented:
Once the glyph subset review is complete, this branch is ready to merge with `dev-build-scripts` from my standpoint.  Let me know if there is anything further that I can do here.

chrissimpkins commented:
@burodepeper any issues with merge of this to `dev-build-scripts`? I added the woff/woff2 subset glyph reviews to the TODO list in #286

burodepeper commented:
@chrissimpkins I haven't had time to proof yet. Have you ever had issues when building the webfonts with missing or visually distorted glyphs? If you haven't, I suggest you merge. If you have, I'll need to find some time to have a look at it.

chrissimpkins commented:
This will be the first time that we have used this subsetting approach to create the web fonts so we have no prior experience with it.  sfnt2woff-zopfli and woff2 are two established tools for these conversions so I am less concerned with the final renders as with the contents of the subsets themselves.  I just haven't had the time to go through them to confirm that we are putting together the glyph sets that we think that we are.  If I get to it before you do I will let you know.

OK with a merge to `dev-build-scripts` for now and close of this PR so that I can continue to base the rest of the build scripting work on the progress here without a new branch?  We have an IR open for the glyph set contents that is part of the v3.0 milestone.  It won't be released until all tasks are complete.  If you happen to decide to tackle this issue, let me know when you are ready and I will confirm that the web font subsets are being built as they were at this stage.  I don't intend to change the subsetting approach in the scripts *unless we find problems with the current approach*.

burodepeper commented:
Of course, merge the hell away ;)

chrissimpkins commented:
:+1:

chrissimpkins commented:
Also, if you haven't been able to set up the development / build tools yet, you can create the web fonts from any local ttf builds with this OS X tool:

https://github.com/source-foundry/Woffle

It uses the same approach that we are using here.

chrissimpkins commented:
merged
